### PR TITLE
[WIP] Implement simple garbage collector

### LINKED ISSUE #165: Identify GC roots on the stack
Based on #149 and #164, we need to build the capability to identify at runtime the GC roots on the stack.

Concretely, we want to add a C file to the runtime directory that provides a `markRoots` function that can be used by our garbage collector:

```
typedef char *Root;
typedef void (*AddRootFunctionPtr)(Root *root, void *arg);
void markRoots(AddRootFunctionPtr func, void *arg);
```

Not sure if this is the best API. We should probably look MMTK and figure out something that is a good fit.

### LINKED ISSUE #300: Implement simple garbage collector
This follows on from #165.

I think the best approach at this stage is a Cheney-style semispace copying collector, which is described here https://en.wikipedia.org/wiki/Cheney%27s_algorithm

It's very elegant and simple.

Concretely:

1. use mmap to allocate the heap on startup
2.  heap will be page aligned
3. I suggest a default size (including both halves) of 2Gb and multiple of page size; it should be possible to override this with an environment variable
4. allocation rounds up to a multiple of 8
5. each allocation starts with a 64-bit word that usually holds the size in bytes, rounded up, but during collection can also hold the forwarding pointer
6. we need to be able to tell when we have a forwarding pointer; since pointers and sizes are multiples of 8, we can just | with 1 for one of the cases
7. when scanning the stack, if there are derived pointers, we need to be able to find the base pointer i.e. in the derived pointer case, the marking function needs to get an extra argument that is the base pointer
8. when it does an allocation, if the heap is full, it will run a collection; if that doesn't free enough space, then it aborts
9. in between collections, we should mprotect (as PROT_NONE) the semispace that is not being used in order to detect bugs

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Implement simple garbage collector.
Resolves #300
Resolved #165

### CODE REVIEW

File: runtime/gc_cheney.c [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+void collect() {
+    printf("%s\n", "Collect");
+    get_roots();
+    // get_roots() -> calls function in gc-roots, should pass rsp as parameter
+    // swap from_space <-> to_space
+    // should update heap_limit_ptr
+}
+
+UntypedPtr _bal_alloc(uint64_t nBytes) {
+    nBytes = nBytes + HEAP_HEADER_SIZE;
```
Comment (jclark) [inline runtime/gc_cheney.c:59]: You need to round up nBytes to a multiple of 8.

File: runtime/gc_cheney.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    to_space_ptr = mmap(NULL, heap_half_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
+    if (to_space_ptr == MAP_FAILED) {
+        printf("Heap initialization Failed\n");
+        abort();
+    }
+}
+
+void collect() {
+    printf("%s\n", "Collect");
+    get_roots();
```
Comment (jclark) [inline runtime/gc_cheney.c:52]: get_roots needs to take a function pointer that it will call for each root found

File: runtime/gc_cheney.c [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+uint8_t *alloc_ptr;
+uint8_t *scan_ptr;
+uint8_t *heap_limit_ptr;
+
+extern void get_roots();
+
+void _bal_init_heap() {
+    int page_size = getpagesize();
+    const char *heap_env = getenv("BAL_HEAP");
+    uint64_t heap_half_size = (heap_env != NULL) ? atoi(heap_env) : DEFAULT_HEAP_HALF_SIZE;
```
Comment (jclark) [inline runtime/gc_cheney.c:30]: Need to check return of atoi

**Review by jclark — APPROVED** [review]
Why do we need libbacktrace?

**Comment by jclark** [comment]
Use however big a heap you need to get the tests to run.  We can reduce when GC is working.

File: runtime/gc_cheney.c (hunk 1 above)
Comment (KavinduZoysa) [reply runtime/gc_cheney.c:59]: Fixed in https://github.com/ballerina-platform/nballerina/commit/8964a5ea73124ee8d2132121b66b78e9f0be39a4

File: runtime/gc_cheney.c (hunk 3 above)
Comment (KavinduZoysa) [reply runtime/gc_cheney.c:30]: Fixed in https://github.com/ballerina-platform/nballerina/commit/8964a5ea73124ee8d2132121b66b78e9f0be39a4

**Comment by KavinduZoysa** [comment]
> Why do we need libbacktrace?

It is used to get the stack trace to find GC roots. https://github.com/ballerina-platform/nballerina/blob/8964a5ea73124ee8d2132121b66b78e9f0be39a4/runtime/gc-roots/roots.c#L50

**Comment by KavinduZoysa** [comment]
> Use however big a heap you need to get the tests to run. We can reduce when GC is working.

Okay, if the heap half size is 1GB, two tests are failing (Vbenchmap and Vbenchmergesort), I have increased it it 3GB. Now all tests are passing.

**Comment by jclark** [comment]
> > Why do we need libbacktrace?
> 
> It is used to get the stack trace to find GC roots.

But I thought statepoint-utils was doing a GC without needing libbacktrace

**Comment by jclark** [comment]
It's not a problem that you are using libbacktrace, I just don't understand.

Also bear in mind that I will want to do #167 soon, which will need libbacktrace. This may happen before you are done with GC.

**Comment by KavinduZoysa** [comment]
> But I thought statepoint-utils was doing a GC without needing libbacktrace

> It's not a problem that you are using libbacktrace, I just don't understand.
> 

The `statepoint-utils` goes through the stack using a method that depends on the architecture, i.e. at the end of each stack frame(stack pointer), it keeps the return address of the previous function. So what `statepoint-utils` does is adding the frame size to stack pointer and get the value of that location. Since this method depends on the architecture, I use libbackatrace to get the return address of each frame.

> Also bear in mind that I will want to do #167 soon, which will need libbacktrace. This may happen before you are done with GC.

Sure, will do

**Pull request merged**